activation/nipost.go

```diff
@@ -222,5 +195,148 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.Hash3
 		NIPost: &types.NIPost{},
 	}
 	nb.persist()
-	return nipost, nil
+	return nipost, postGenDuration, nil
+}
+
+// Submit the challenge to a single PoET.
+func submitPoetChallenge(ctx context.Context, logger log.Log, poet PoetProvingServiceClient, challenge *types.Hash32) (*PoetRequest, error) {
+	poetServiceID, err := poet.PoetServiceID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get PoET service ID: %v", ErrPoetServiceUnstable, err)
+	}
+
+	logger.With().Debug("submitting challenge to poet proving service",
+		log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+		log.Stringer("challenge", *challenge))
+
+	round, err := poet.Submit(ctx, *challenge)
+	if err != nil {
+		logger.With().Error("failed to submit challenge to poet proving service",
+			log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+			log.Stringer("challenge", *challenge),
+			log.Err(err))
+		return nil, fmt.Errorf("%w: failed to submit challenge to poet service: %v", ErrPoetServiceUnstable, err)
+	}
+
+	logger.With().Info("challenge submitted to poet proving service",
+		log.String("poet_id", util.Bytes2Hex(poetServiceID)),
+		log.String("round_id", round.ID),
+		log.Stringer("challenge", *challenge))
+
+	return &PoetRequest{
+		PoetRound:     round,
+		PoetServiceID: poetServiceID,
+	}, nil
+}
+
+// Submit the challenge to all registered PoETs.
+func (nb *NIPostBuilder) submitPoetChallenges(ctx context.Context, challenge *types.Hash32) []PoetRequest {
+	g, ctx := errgroup.WithContext(ctx)
+	poetRequestsChannel := make(chan PoetRequest)
+	for _, poetProver := range nb.poetProvers {
+		poet := poetProver
+		g.Go(func() error {
+			if poetRequest, err := submitPoetChallenge(ctx, nb.log, poet, challenge); err == nil {
+				poetRequestsChannel <- *poetRequest
+			} else {
+				nb.log.With().Warning("failed to submit challenge to PoET", log.Err(err))
+			}
+			return nil
+		})
+	}
+	poetRequests := make([]PoetRequest, 0)
+	done := make(chan struct{})
+	go func() {
+		for request := range poetRequestsChannel {
+			poetRequests = append(poetRequests, request)
+		}
+		close(done)
+	}()
```

Comment:
i think this is cleaner
```
requests := make(chan Request, N)
eg.Go // spawn N goroutines
g.Wait() // because every goroutine in eg.Go wont block on send to requests
close(requests)
... loop over closed channel ...
```